src/useMatches.tsx

```diff
@@ -56,25 +75,70 @@ export function useMatches() {
   const matches = useInternalMatches(filtered, search);
 
   const results = React.useMemo(() => {
-    let groupMap: Record<string, ActionImpl[]> = {};
+    /**
+     * Store a reference to a section and it's list of actions.
+     * Alongside these actions, we'll keep a temporary record of the
+     * final priority calculated by taking the commandScore + the
+     * explicitly set `action.priority` value.
+     */
+    let map: Record<SectionName, { priority: number; action: ActionImpl }[]> =
+      {};
+    /**
+     * Store another reference to a list of sections alongside
+     * the section's final priority, calculated the same as above.
+     */
+    let list: { priority: number; name: SectionName }[] = [];
+    /**
+     * We'll take the list above and sort by its priority. Then we'll
+     * collect all actions from the map above for this specific name and
+     * sort by its priority as well.
+     */
+    let ordered: { name: SectionName; actions: ActionImpl[] }[] = [];
+
     for (let i = 0; i < matches.length; i++) {
-      const action = matches[i];
-      const section = action.section || NO_GROUP;
-      if (!groupMap[section]) {
-        groupMap[section] = [];
+      const match = matches[i];
+      const action = match.action;
+      const score = match.score || Priority.NORMAL;
+
+      const section = {
+        name:
+          typeof action.section === "string"
+            ? action.section
+            : action.section?.name || NO_GROUP.name,
+        priority:
+          typeof action.section === "string"
+            ? score
+            : action.section?.priority || 0 + score,
+      };
+
+      if (!map[section.name]) {
+        map[section.name] = [];
+        list.push(section);
```

Comment:
The downside of allowing the same `section` defined on multiple actions is that on one action, `section` may just be a string; `"Settings"`. On another, the `section` could be:

```
{
  name: "Settings",
  priority: Priority.HIGH
}
```

With the current implementation, the first action's `section` that appears at runtime will be the value we use.